bxsolana/provider/ws.ts

```diff
@@ -60,229 +60,428 @@ import {
+            this.streamToCountMap.set("GetOrderbooksStream", countToSubscriptionID)
+        } else {
+            count = this.streamToCountMap.size + 1
+            const countToSubscriptionID = this.streamToCountMap.get("GetOrderbooksStream")
+
+            if (countToSubscriptionID) {
+                countToSubscriptionID.set(count, subscriptionId)
+                this.streamToCountMap.set("GetOrderbooksStream", countToSubscriptionID)
+            }
+        }
+
+        return this.wsConnection.subscribeToNotifications(subscriptionId)
     }
 
-    getTickersStream(request: GetTickersRequest): Promise<AsyncGenerator<GetTickersStreamResponse>> {
-        return this.wsSocketStreamCall("GetTickersStream", request)
+    // Since the implementation of the subscription functions rely on the count incrementing on every subsequent call
+    // to the same stream, the cancel functionality operates on the assumption that the "streamNumber" corresponds to
+    // the specific call of the stream by the user
+
+    // example
+    // getOrderbooksStream (count === 1)
+    // getOrderbooksStream (count === 2)
+
+    // can cancel 1, or 2. If a nonvalid cancellation number is sent as an input, the promise will be rejected with a
+    // false boolean
+
+    cancelGetOrderbooksStreamByCount = async (streamNumber: number): Promise<boolean> => {
+        const countToSubscriptionID = this.streamToCountMap.get("GetOrderbooksStream")
+
+        if (countToSubscriptionID) {
+            this.countToSubscriptionID.forEach((value: string, key: number) => {
+                if (key === streamNumber) return this.wsConnection.unsubscribe(value)
+            })
+        }
+
+        return false
     }
 
-    getTradesStream(request: GetTradesRequest): Promise<AsyncGenerator<GetTradesStreamResponse>> {
-        return this.wsSocketStreamCall("GetTradesStream", request)
+    cancelAllGetOrderbooksStream = async (): Promise<boolean> => {
+        let retValues: Promise<boolean>
+        let p = Promise.resolve(false)
+
+        const countToSubscriptionID = this.streamToCountMap.get("GetOrderbooksStream")
+
+        if (countToSubscriptionID) {
+            countToSubscriptionID.forEach((value: string, key: number) => {
+                retValues = this.wsConnection.unsubscribe(value)
+                retValues.catch(function () {
+                    return retValues
+                })
+            })
+
+            return p
+        }
+
+        p = Promise.reject(true)
+        return p
     }
 
-    getOrderStatusStream(request: GetOrderStatusStreamRequest): Promise<AsyncGenerator<GetOrderStatusStreamResponse>> {
-        return this.wsSocketStreamCall("GetOrderStatusStream", request)
+    getTickersStream = async (request: GetTickersRequest): Promise<AsyncGenerator<GetTickersStreamResponse>> => {
+        const subscriptionId = await this.wsConnection.subscribe("GetTickersStream", request)
+
+        this.streamMap.set("GetTickersStream", subscriptionId)
```

Comment:
Doesn't this need to have the same logic as `getOrderbookStream`, with all the counters and such? Looks to me like it does and you'll need to do a bit of refactoring to avoid copying code in all of these streams.